File: backend/enum-jpa-validation-app/src/main/java/com/example/enumapp/web/ApiResponse.java
package com.example.enumapp.web;

import com.fasterxml.jackson.annotation.JsonInclude;

/**
 * 성공/실패 공통 API 응답 구조.
 * <pre>
 * 성공:
 * {
 *   "status": 200,
 *   "error": false,
 *   "result": { ... }
 * }
 *
 * 실패:
 * {
 *   "status": 400,
 *   "error": true,
 *   "result": {
 *     "message": "...",
 *     "messageCode": "VALIDATION_ERROR",
 *     "errors": { "field": "..." }
 *   }
 * }
 * </pre>
 */
@JsonInclude(JsonInclude.Include.ALWAYS)
public class ApiResponse<T> {

    private int status;
    private boolean error;
    private T result;

    public static <T> ApiResponse<T> ok(int status, T result) {
        ApiResponse<T> response = new ApiResponse<>();
        response.status = status;
        response.error = false;
        response.result = result;
        return response;
    }

    public static ApiResponse<ApiErrorBody> fail(int status, String messageCode, String message) {
        return fail(status, messageCode, message, null);
    }

    public static ApiResponse<ApiErrorBody> fail(
            int status,
            String messageCode,
            String message,
            java.util.Map<String, String> errors
    ) {
        ApiResponse<ApiErrorBody> response = new ApiResponse<>();
        response.status = status;
        response.error = true;
        response.result = ApiErrorBody.of(messageCode, message, errors);
        return response;
    }

    public int getStatus() {
        return status;
    }

    public boolean isError() {
        return error;
    }

    public T getResult() {
        return result;
    }
}

File: enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/web/ApiResponse.kt
package com.example.enumapp.web

import com.fasterxml.jackson.annotation.JsonInclude

/**
 * 성공/실패 공통 API 응답 구조.
 * <pre>
 * 성공:
 * {
 *   "status": 200,
 *   "error": false,
 *   "result": { ... }
 * }
 *
 * 실패:
 * {
 *   "status": 400,
 *   "error": true,
 *   "result": {
 *     "message": "...",
 *     "messageCode": "VALIDATION_ERROR",
 *     "errors": { "field": "..." }
 *   }
 * }
 * </pre>
 */
@JsonInclude(JsonInclude.Include.ALWAYS)
class ApiResponse<T> private constructor(
    val status: Int,
    val error: Boolean,
    val result: T?,
) {
    companion object {
        fun <T> ok(status: Int, result: T?): ApiResponse<T> =
            ApiResponse(status, false, result)

        fun fail(status: Int, messageCode: String?, message: String?): ApiResponse<ApiErrorBody> =
            fail(status, messageCode, message, null)

        fun fail(
            status: Int,
            messageCode: String?,
            message: String?,
            errors: Map<String, String>?,
        ): ApiResponse<ApiErrorBody> =
            ApiResponse(status, true, ApiErrorBody.of(messageCode, message, errors))
    }
}
